guides/hack/03-expressions-and-operators/27-function-call-examples/array-of-closures.php

```diff
@@ -17,10 +17,27 @@ function(int $p) {
     ), // times 5
   ];
 
-  $i = 0;
-  $v = $table[$i++]($i); // eval is L-to-R
+  $incr = new Incrementer();
+
+  $v = $table[$incr->postIncr()]($incr->value()); // eval is L-to-R
   echo "\$v = $v\n";
 
-  $v = $table[$i](++$i); // eval is L-to-R
+  $v = $table[$incr->value()]($incr->preIncr()); // eval is L-to-R
   echo "\$v = $v\n";
 }
+
+class Incrementer {
```

Comment:
I don't have a strong opinion but given that this is meant to be a relatively minor example in the middle of a larger guide (https://docs.hhvm.com/hack/expressions-and-operators/function-call), I think it might be preferable not to make it even longer than it already is...

In my attempt (#778), I basically removed that part of the example completely which I think is fine, but it might also be fine to keep it if it doesn't make it much longer.

Maybe something like this?

```hack
$v = $table[next_number()](next_number());
echo "\$v = $v\n";

$v = $table[next_number()](next_number());
echo "\$v = $v\n";

...

function next_number(): Generator<int, int, nothing> {
  for ($i = 0; true; $i++) {
    yield $i;
  }
}
```

The pre/post-increment difference doesn't seem to be relevant for demonstrating the L-to-R behavior.

Reply:
I might be misunderstanding how this program works. Vecs are not indexable with generators. I see what you want here (a function that returns n + 1 where n is the amount of times the function has been called before - 1).

Doing that requires either:
 - Calling `next()` and `current()` on the generator, which can't be done.
 - A lambda capturing a `HH\Lib\Ref<int>` and using that as storage.
 - A class with a private static member to keep the current value in.
 - A static variable in the function (but that died in an hhvm version a long time ago).

I think I'll go with:
```HACK
/**
 * Returns a function which will return the next
 * number in a sequence, starting at 0, 1, 2, 3, etc.
 */
function sequence(): (function(): int) {
  $ticket = new Ref(-1);
  return () ==> {
    $ticket->value++;
    return $ticket->value;
  };
}
```

Reply:
Oh of course, I don't know what I was thinking. I guess a class (like you originally had) is really the simplest working example then, maybe just a simpler class, something like:

```hack
class Incrementer {
  private int $last = -1;
  public function next(): int {
    $this->last++;
    return $this->last;
  }
}
```

Reply:
(I'd guess that most readers probably have better understanding of classes than of higher order functions.)

Reply:
Yeah, now I don't retain the `$i++`, `$i`, and `++$i` semantics, this would do best.